src/par/FixColumnPartitions.c

```diff
@@ -52,14 +53,18 @@ int FixColumnPartitions_IsSideFace(Mesh_ptr mesh, MRegion_ptr mr, MFace_ptr rf)
    zero volume and is therefore uncertain.
 */
 int FixColumnPartitions_UpDownFaces(Mesh_ptr mesh, MRegion_ptr mr, MFace_ptr* up, MFace_ptr* dn) {
-  int found;
-  List_ptr rfaces, rfaces2;
+  int found, up_unknown;
+  List_ptr rfaces, fregs;
   int nrf, i, j, k, nfv, ret;
-  MFace_ptr rf; double rcen[3], upcen[3], dncen[3];
+  MFace_ptr rf, curf_it, nxtf_it;
+  double upcen[3], dncen[3];
+  double cen_it_z;
   double fxyz[MAXPV2][3];
   MRegion_ptr mr_it;
 
-  *up=NULL; *dn=NULL;
+  *dn=NULL;
+  if (*up == NULL) up_unknown = 1;
```

Comment:
There are two scenarios under which we call the function to find top&bottom faces: in one case, we know nothing (when originally looking for the top face in the column) and in the other we know the top face and look for the bottom one (when going top down in the column). In the first case, it's hard to avoid geometrical checks unless you want to do a lot of searching. In the other case, you can just use the topology. So, if the passed *up==NULL, it will use geometrical checks and look for both top and bottom. If it's given, then it will pick the other non-side face as the *dn.